### PR TITLE
feat: Allow Users to Specify Kubernetes secret for TLS. Fixes #5582

### LINKED ISSUE #5582: TLS configuration and/or docs enhancement
# Summary

It looks like after 3.0 release argo server listens on https port with self-signed certificate (#5581 )
I think it would be great to have a dedicated ConfigMap with tls.cert and tls.key parameters which would be mounted inside pod (Like it is implemented in ArgoCD) so it would be very easy to generate a valid certificate with cert-manager or ue any pre-isued one.
At least it would be great to have a parameter or doc which would tell where the cert is read from. It looks like it is harcoded now and it's impossible to mount CM inside pod without even knowing where the self-signed cert is located...

# Use Cases

When deploying a new instance with Ingress it would be great to use ACME from cert-manager

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by ChaosInTheCRD**
Merge After: #6540

This PR allows a user to specify a secret name as a command-line parameter (`--tls-certificate-secret-name`) on the argo-server. Said secret would contain keys `tls.key` and `tls.crt` in order to provide a mechanism by which they can bring their own certificates (e.g. from cert-manager, Bitnami Sealed Secrets, generic Kubernetes secret).

This PR tries to follow a similar implementation to what is seen in Argo CD (https://github.com/argoproj/argo-cd/pull/6071), for maintainability.

Marked as WIP until the following is completed:
- Unit Tests
- Documentation

### CODE REVIEW

**Comment by alexec** [comment]
Change to draft while WIP.

**Review by alexec — CHANGES_REQUESTED** [review]
Needs sync with master

**Comment by ChaosInTheCRD** [comment]
@alexec I am looking to take some time, hopefully tomorrow, to get this ready for release. I'll sync it with master, make sure unit tests are written and update documentation.

If there's anything else you need me to do please let me know.

Thanks!

File: cmd/argo/commands/server.go [hunk 1]
```diff
@@ -104,21 +105,28 @@ See %s`, help.ArgoServer),
...
 			if secure {
-				log.Infof("Generating Self Signed TLS Certificates for Secure Mode")
 				tlsMinVersion, err := env.GetInt("TLS_MIN_VERSION", tls.VersionTLS12)
 				if err != nil {
 					return err
 				}
-				var cer *tls.Certificate
-				cer, err = tlsutils.GenerateX509KeyPair()
+
+				if tlsCertificateSecretName != "" {
```
[the remark this answers is not in the document]
Comment (alexec) [reply cmd/argo/commands/server.go:113]: I think this works well when running locally vs in container.

**Review by alexec — CHANGES_REQUESTED** [review]
Great stuff. Please:

* [ ] Sync with master.
* [ ] Run `make pre-commit -B` and push changes.
* [ ] Dismiss reviews.

**Review by alexec — COMMENTED** [review]
is this ready for review again? need any help?

File: cmd/argo/commands/server.go [hunk 2]
```diff
@@ -105,11 +108,30 @@ See %s`, help.ArgoServer),
 
 			var tlsConfig *tls.Config
 			if secure {
-				log.Infof("Generating Self Signed TLS Certificates for Secure Mode")
-				tlsConfig, err = tlsutils.GenerateTLSConfig()
+				tlsMinVersion := env.GetInt("TLS_MIN_VERSION", tls.VersionTLS12)
```
Comment (alexec) [inline cmd/argo/commands/server.go:111]: I'm pretty sure there is existing func for this. Adding a new go.mod dependency increases our vulnerability surface, so please remove this.

**Comment by ChaosInTheCRD** [comment]
@alexec sorry for not updating this PR in so long. Unfortunately, the state of the commits and the fact that so many of them are signed has held me back. I will fix this now.

**Comment by ChaosInTheCRD** [comment]
closing this PR in place of https://github.com/argoproj/argo-workflows/pull/7621

**Pull request closed without merging**